streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -325,6 +334,12 @@ public void checkpoint(final Map<TopicPartition, Long> ackedOffsets) {
         }
         // write the checkpoint file before closing, to indicate clean shutdown
         try {
+            if (checkpoint == null) {
+                assert eosEnabled;
+                checkpoint = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
+            } else {
+                assert !eosEnabled;
```

Comment:
If you hit a method 1M per second because it's on the hot code path, it would have performance impact... IMHO, that's exactly the difference between exceptions vs assertions. If you assert _everywhere_ you don't want to have this in production. And all checks that do make sense in production should be exceptions :)

But it might be a philosophical question.